LIPS/lip-25.md

```diff
@@ -0,0 +1,1983 @@
+
+```solidity
+error ZeroAddressLido();
+error ZeroAddressAdmin();
+
+/**
+* @dev proxy initialization
+* @param _admin Lido DAO Aragon agent contract address
+* @param _lido Lido address
+* @param _withdrawalCredentials Lido withdrawal vault contract address
+*/
+function initialize(address _admin, address _lido, bytes32 _withdrawalCredentials) external {
+    if (_admin == address(0)) revert ZeroAddressAdmin();
+    if (_lido == address(0)) revert ZeroAddressLido();
+
+    _initializeContractVersionTo(2);
+
+    _setupRole(DEFAULT_ADMIN_ROLE, _admin);
+
+    LIDO_POSITION.setStorageAddress(_lido);
+    WITHDRAWAL_CREDENTIALS_POSITION.setStorageBytes32(_withdrawalCredentials);
+    emit WithdrawalCredentialsSet(_withdrawalCredentials, msg.sender);
+}
+
+/**
+* @notice A function to finalize upgrade to v2 (from v1). Can be called only once
+* @param _priorityExitShareThresholds array of priority exit share thresholds
+* @param _maxDepositsPerBlock array of max deposits per block
+* @param _minDepositBlockDistances array of min deposit block distances
+*/
+function finalizeUpgrade_v2(
+    uint256[] memory _priorityExitShareThresholds,
+    uint256[] memory _maxDepositsPerBlock,
+    uint256[] memory _minDepositBlockDistances
+) external {
+    _checkContractVersion(1);
+
+    uint256 stakingModulesCount = getStakingModulesCount();
+
+    if (stakingModulesCount != _priorityExitShareThresholds.length) {
+        revert ArraysLengthMismatch(stakingModulesCount, _priorityExitShareThresholds.length);
+    }
+
+    if (stakingModulesCount != _maxDepositsPerBlock.length) {
+        revert ArraysLengthMismatch(stakingModulesCount, _maxDepositsPerBlock.length);
+    }
+
+    if (stakingModulesCount != _minDepositBlockDistances.length) {
+        revert ArraysLengthMismatch(stakingModulesCount, _minDepositBlockDistances.length);
+    }
+
+    for (uint256 i; i < stakingModulesCount; ) {
+        StakingModule storage stakingModule = _getStakingModuleByIndex(i);
+        _updateStakingModule(
+            stakingModule,
+            stakingModule.id,
+            stakingModule.stakeShareLimit,
+            _priorityExitShareThresholds[i],
+            stakingModule.stakingModuleFee,
+            stakingModule.treasuryFee,
+            _maxDepositsPerBlock[i],
+            _minDepositBlockDistances[i]
+        );
+
+        unchecked {
+            ++i;
+        }
+    }
+
+    _updateContractVersion(2);
+}
+```
+
+#### 2.3. Keys vetting improvements
+
+Support of new key vetting logic requires implementation of the new `decreaseStakingModuleVettedKeysCountByNodeOperator` external method. Also, the new `STAKING_MODULE_UNVETTING_ROLE` constant should be defined.
```

Comment:
> Also, the new `STAKING_MODULE_UNVETTING_ROLE` constant should be defined.

Basically what we're proposing here is to add `STAKING_MODULE_UNVETTING_ROLE` role and assign it to a DSM

Reply:
OK, added this note to the 2.3 section.